dss/operations/storage.py

```diff
@@ -0,0 +1,128 @@
+
+from dss import Config, Replica
+from dss.storage.hcablobstore import compose_blob_key
+from dss.operations import dispatch
+from dss.operations.util import CommandForwarder, map_bucket
+from dss.events.handlers.sync import dependencies_exist
+
+
+logger = logging.getLogger(__name__)
+
+
+class StorageOperationHandler:
+    def __init__(self, argv: typing.List[str], args: argparse.Namespace):
+        self.keys = args.keys.copy() if args.keys else None
+        self.entity_type = args.entity_type
+        self.job_id = args.job_id
+        if args.replica:
+            self.replica = Replica[args.replica]
+            self.handle = Config.get_blobstore_handle(self.replica)
+
+    def forward_command_to_lambda(self, argv: typing.List[str], args: argparse.Namespace):
+        """
+        This transforms a command into a format appropriate for Lambda execution. To take advantage of Lambda scaling,
+        commands operating any multiple keys are forwarded as multiple commands operating on a single key.
+        """
+        cmd_template = f"{argv[0]} {argv[1]}"
+        if "entity_type" in args.__dict__.keys() and "keys" in args.__dict__.keys():
+            del args.entity_type
+        for argname, argval in args.__dict__.items():
+            argname = argname.replace("_", "-")
+            if argname not in ["forward-to-lambda", "keys", "func"]:
+                cmd_template += f" --{argname} {argval}"
+        cmd_template += " --keys {}"
+
+        def forward_keys(keys):
+            with CommandForwarder() as f:
+                for key in keys:
+                    f.forward(cmd_template.format(key))
+
+        if args.keys is not None:
+            forward_keys(args.keys)
+        else:
+            map_bucket(forward_keys, self.handle, self.replica.bucket, f"{self.entity_type}s/")
+
+    def process_command_locally(self, argv: typing.List[str], args: argparse.Namespace):
+        if self.keys is not None:
+            for key in self.keys:
+                self.process_key(key)
+        else:
+            def process_keys(keys):
+                for key in keys:
+                    self.process_key(key)
+
+            map_bucket(process_keys, self.handle, self.replica.bucket, f"{self.entity_type}s/")
+
+    def log_warning(self, name: str, info: dict):
+        logger.warning(json.dumps({'job_id': self.job_id, name: info}))
+
+    def process_key(self, key):
+        raise NotImplementedError()
+
+    def __call__(self, argv: typing.List[str], args: argparse.Namespace):
+        if args.forward_to_lambda:
+            self.forward_command_to_lambda(argv, args)
+        else:
+            self.process_command_locally(argv, args)
+
+storage = dispatch.target(
+    "storage",
+    arguments={"--forward-to-lambda": dict(default=False, action="store_true"),
+               "--replica": dict(choices=[r.name for r in Replica], required=True),
+               "--entity-type": dict(choices=["file", "bundle", "collection"]),
```

Comment:
Should these be pulled from `https://github.com/HumanCellAtlas/data-store/blob/master/dss/storage/identifiers.py` since they're being used as the prefix identifiers for bucket keys?

Reply:
That's reasonable, done!